### PR TITLE
Pandas furness wrapper

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Is the documentation building correctly with a clean version number?
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by MattBuckley-TfN** [comment]
@isaac-tfn  what is left to do on this PR before it's ready for review?

**Comment by isaac-tfn** [comment]
> @isaac-tfn what is left to do on this PR before it's ready for review?

Probably just the final linting and stuff. Probably should write tests too but it may be better just to get it published

File: requirements.txt [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 tqdm>=4.50.2
-caf.toolkit>=0.7.4
+caf.toolkit>=0.11
 matplotlib>=3.8.2
+xarray>=2024
 
 # Ipf requirements
-sparse>=0.13.0
+sparse>=0.13.0, <0.16
 numba>=0.60.0
```
Comment (BenTaylor-TfN) [inline requirements.txt]: Can these be removed from here as they are only needed from the sparse optional?

```suggestion
```

File: requirements_sparse.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+sparse>=0.15.4
```
Comment (BenTaylor-TfN) [inline requirements_sparse.txt]: Moving in `numba` from comment above.

```suggestion
sparse>=0.15.4
numba>=0.60.0
```

Comment (BenTaylor-TfN) [reply requirements_sparse.txt]: Does numba still need a cap on numpy? If tests are passing atm I guess it's a future problem?

File: pyproject.toml [hunk 3]
```diff
@@ -97,11 +76,11 @@ import_heading_firstparty = "Local Imports"
 
 # Pylint settings
 [tool.pylint.basic]
-good-names = ["df", "fn", "ca", "tp", "to", "x", "x1", "x2", "i", "f", "q", "mu"]
+good-names = ["df", "fn", "ca", "tp", "to", "x", "x1", "x2", "i", "f", "q"]
 extension-pkg-allow-list = ["pydantic"]
 
 [tool.pylint.main]
-ignore = ["__init__.py", "_version.py"]
+ignore = ["__init__.py", "_version.py", "versioneer.py"]
```
Comment (BenTaylor-TfN) [inline pyproject.toml]: Shouldn't be needed anymore as we don't use versioneer
```suggestion
ignore = ["__init__.py", "_version.py"]
```

**Review by BenTaylor-TfN — COMMENTED** [review]
I've only reviewed files that are **not** in `src`. Would be better if someone who works with this code more regularly reviews the code content.

File: requirements.txt [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
+pandas>=2
 tqdm>=4.50.2
-caf.toolkit>=0.7.4
+caf.toolkit>=0.11
 matplotlib>=3.8.2
+xarray>=2024
 
 # Ipf requirements
-sparse>=0.13.0
-numba>=0.60.0
```
Comment (MattBuckley-TfN) [inline requirements.txt]: Unnecessary comment
```suggestion
```

File: pyproject.toml [hunk 5]
```diff
@@ -4,56 +4,33 @@ build-backend = "setuptools.build_meta"
...
-    "pytest-xdist>=3.3.1",
-]
-
-[project.urls]
-"Bug Tracker" = "https://github.com/Transport-for-the-North/caf.distribute/issues"
-Homepage = "https://github.com/Transport-for-the-North/caf.distribute"
-Source = "https://github.com/Transport-for-the-North/caf.distribute"
-Documentation = "http://cafdistribute.rtfd.io/"
+# Dynamic versioning for editable installs
+[tool.setuptools.cmdclass]
```
Comment (MattBuckley-TfN) [inline pyproject.toml:31]: Now this is here I think setup.cfg can be removed

File: src/caf/distribute/gravity_model/core.py [hunk 6]
```diff
@@ -187,8 +189,9 @@ def _tidy_unique_id(unique_id: str) -> str:
     @property
     def achieved_band_share(self) -> np.ndarray:
         """The achieved band share values during the last run."""
-        if self.achieved_cost_dist is None:
+        if not isinstance(self.achieved_cost_dist, cost_utils.CostDistribution):
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/core.py:194]: Should there be a different error if this is a list of distributions, or can it just work with a list of distributions?

File: src/caf/distribute/gravity_model/multi_area.py [hunk 7]
```diff
@@ -384,7 +384,7 @@ class MGMCostDistribution:
     @classmethod
     def from_pandas(
         cls,
-        category: str | int,
+        category: str,
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:387]: docstring type hasn't been updated

File: src/caf/distribute/gravity_model/multi_area.py [hunk 8]
```diff
@@ -586,7 +586,7 @@ def calibrate(
         gm_params: GMCalibParams,
         verbose: int = 0,
         **kwargs,
-    ) -> dict[str | int, GravityModelResults]:
+    ) -> dict[str, GravityModelResults]:
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:589]: docstring type hasn't been updated

File: src/caf/distribute/gravity_model/multi_area.py [hunk 9]
```diff
@@ -878,7 +883,7 @@ def run(
         distributions: MultiCostDistribution,
         running_log_path: Path,
         furness_tol: float = 1e-6,
-    ) -> dict[int | str, GravityModelResults]:
+    ) -> dict[str, GravityModelResults]:
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:886]: docstring return type hasn't been updated

File: src/caf/distribute/gravity_model/single_area.py [hunk 10]
```diff
@@ -114,9 +114,9 @@ def _gravity_function(
         self._loop_start_time = timing.current_milli_time()
 
         # Update performance params
-        self.achieved_cost_dist = cost_distribution
-        self.achieved_convergence = convergence
-        self.achieved_distribution = matrix
+        self.achieved_cost_dist: cost_utils.CostDistribution = cost_distribution
+        self.achieved_convergence: float = convergence
+        self.achieved_distribution: np.ndarray = matrix
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/single_area.py:119]: Instance variables should all be defined within the `__init__` even if that just means they're set to none. 

I don't think this is a problem that needs fixing now because it looks like there's quite alot of them but might be worth tidying up in future to stop pylint complaining.

File: src/caf/distribute/furness.py [hunk 11]
```diff
@@ -157,3 +161,234 @@ def doubly_constrained_furness(
...
+    max_iters: int = 2000,
+    seed_infill: float = 1e-3,
+    normalise_seeds: bool = True,
+    tol: float = 1e-9,
+    idx_col: str = "model_zone_id",
+    unique_col: str = "trips",
+    round_dp: int = 8,
+    unique_zones: list[int] | None = None,
+    unique_zones_join_fn: Callable = operator.and_,
+) -> tuple[pd.DataFrame, int, float]:
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/furness.py:181]: Pylint isn't happy about too-many-positional-arguments and I would agree, I think just switching all the optional ones to keyword only would be good.
```suggestion
def furness_pandas_wrapper(
    seed_values: pd.DataFrame,
    row_targets: pd.DataFrame,
    col_targets: pd.DataFrame,
    *,
    max_iters: int = 2000,
    seed_infill: float = 1e-3,
    normalise_seeds: bool = True,
    tol: float = 1e-9,
    idx_col: str = "model_zone_id",
    unique_col: str = "trips",
    round_dp: int = 8,
    unique_zones: list[int] | None = None,
    unique_zones_join_fn: Callable = operator.and_,
) -> tuple[pd.DataFrame, int, float]:
```

Above is a breaking change so might break wherever this function is called depending if they used positional arguments for everything.

File: src/caf/distribute/furness.py [hunk 12]
```diff
@@ -157,3 +161,234 @@ def doubly_constrained_furness(
...
+    if len(ref_index.difference(col_targets.index)) > 0:
+        raise ValueError("Row and Column target indexes do not match.")
+
+    if len(ref_index.difference(seed_values.index)) > 0:
+        raise ValueError("Row and Column target indexes do not match seed index.")
+
+    if len(ref_index.difference(seed_values.columns)) > 0:
+        raise ValueError("Row and Column target indexes do not match seed columns.")
+
+    assert_approx_equal(
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/furness.py:274]: `np.allclose` should be used instead of the testing and a ValueError should be raised because assert statements are ignored in Python's optimize mode

File: src/caf/distribute/furness.py [hunk 13]
```diff
@@ -157,3 +161,234 @@ def doubly_constrained_furness(
...
+            df=seed_values,
+            select=unique_zones,
+            join_fn=unique_zones_join_fn,
+        )
+        seed_values = seed_values.where(mask, 0)
+
+    # ## CONVERT TO NUMPY AND FURNESS ## #
+    row_targets = row_targets.values.flatten()
+    col_targets = col_targets.values.flatten()
+    seed_values = seed_values.values
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/furness.py]: Use `to_numpy()` instead of values, I think pandas might have plans to change values in future
```suggestion
    row_targets = row_targets.to_numpy().flatten()
    col_targets = col_targets.to_numpy().flatten()
    seed_values = seed_values.to_numpy()
```

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Logic all looks good just a few minor things:
- Some docstrings need updating after type changes
- Using `to_numpy` instead of `.values` in a few places (pandas recommends)
- Delete `setup.cfg`

File: src/caf/distribute/gravity_model/core.py (hunk 6 above)
Comment (isaac-tfn) [reply src/caf/distribute/gravity_model/core.py:194]: It doesn't work with a list. It is called on members of the list in the multi-tld case

Comment (MattBuckley-TfN) [reply src/caf/distribute/gravity_model/core.py:194]: It might be worth having different error messages to explain that e.g if None then gravity model hasn't run but if list then something else

**Review by MattBuckley-TfN — APPROVED** [review]
LGTM
